### ISSUE TITLE
Improve translation of short sentences

### ISSUE THREAD

**Issue by eu9ene** [mozilla/firefox-translations-training#215]
We have issues with translating shorter sentences and single words. 

What might help:
- Better data cleaning to avoid removing short translations from the training datasets
- Try to find more datasets that contain short sentences

**Comment by gregtatum** [comment]
Or maybe even synthesize it with the alignment data.

**Comment by gregtatum** [comment]
This behavior is also visible with numbers. A good example is to do a list of numbers.

**Comment by gregtatum** [comment]
Verify the fix with:

https://bugzilla.mozilla.org/show_bug.cgi?id=1888972

**Comment by gregtatum** [comment]
Here is a word count distribution for the merged corpus sl-en: https://firefox-ci-tc.services.mozilla.com/tasks/groups/PPCzZRHaTT6Ys4BIhPGT5w

![word count distribution "en"](https://github.com/user-attachments/assets/12c42a05-5fdc-4963-9b36-c887984edeae)

![word count distribution "sl"](https://github.com/user-attachments/assets/d0eb3c3b-daf3-4ed5-8e67-fd1052a06a9e)

Generated via:

```
python3 pipeline/data/analyze.py --file_location https://firefox-ci-tc.services.mozilla.com/api/queue/v1/task/VK5zmxJRTLy0y0WBQ0DRJg/artifacts/public/build/corpus.en.zst --output data --dataset OPUS_corpus --language en

python3 pipeline/data/analyze.py --file_location https://firefox-ci-tc.services.mozilla.com/api/queue/v1/task/VK5zmxJRTLy0y0WBQ0DRJg/artifacts/public/build/corpus.sl.zst --output data --dataset OPUS_corpus --language sl
```

**Comment by marco-c** [comment]
So we have basically 0 sentences with 1 word?

**Comment by gregtatum** [comment]
I filed #878 which suggests augmenting with statistically synthesized single word translations.

**Comment by gregtatum** [comment]
I filed #879 which suggests harvesting short sentences from parallel datasets.

**Comment by ZJaume** [comment]
Here are some stats extracted form the merged corpus for en-ar.

![v1 ar distribution-words](https://github.com/user-attachments/assets/25ddc705-4c47-4309-947b-861a62e398ea)

![v1 en distribution-words](https://github.com/user-attachments/assets/a432627b-286d-4f2b-8ce6-734de0ae2db0)

**Comment by eu9ene** [comment]
@ZJaume see also this feedback https://github.com/mozilla/firefox-translations-models/issues/210. Dutch is a good candidate to test.

**Comment by ZJaume** [comment]
Trying to spot the models affected by the issue using alignments. I translated the 200 shortest sentence pairs from the [Tatoeba Translation Challenge test](https://github.com/Helsinki-NLP/Tatoeba-Challenge) and computed alignments between the source sentences and the reference sententes and between source sentences and translation hypotheses. Then counted the number of tokens that were aligned in each case, in order to see if the translation is producing more words than they should be (therefore the alignments won't, hopefully, align the extra tokens). I've also added the total number of unaligned punctuation tokens because one of the common patterns for affected models seems to be adding extra '-' like it was a sentence from OpenSubtitles.

The columns that we should be looking at are the third and the sixth column (sorry about the sixth, the signs are reversed). The third would mean that, on average, the model is having X% of the tokens being unaligned per sentence, compared to the reference.

From manual inspection, I would say anything above 8 or 9 in the third column might be severely affected and anything above 3 is, at least, worth looking at the output.

One example of what retraining from scratch for a language could look is Dutch. In the table I'm including the newly trained student and it is now achieving very good aligned token percentages.

| lang | avg % aligned tokens ref | avg % aligned tokens hyp | avg % diff | total unaligned punct tokens ref | total unaligned punct tokens hyp | diff |
| --- | -------: | -------: | -----: | -----: | -----: | ----: |
| ar-en | 84.1 | 91.0 | -6.9 | 129 | 64 | 65 |
| bg-en | 85.1 | 73.2 | 11.9 | 127 | 275 | -148 |
| ca-en | 86.9 | 89.9 | -2.9 | 106 | 81 | 25 |
| cs-en | 79.0 | 82.7 | -3.7 | 180 | 149 | 31 |
| da-en | 91.4 | 93.0 | -1.6 | 71 | 60 | 11 |
| de-en | 91.4 | 97.3 | -5.9 | 71 | 20 | 51 |
| el-en | 90.9 | 86.5 | 4.4 | 68 | 103 | -35 |
| en-ar | 99.0 | 97.4 | 1.6 | 4 | 15 | -11 |
| en-bg | 98.2 | 81.0 | 17.2 | 10 | 146 | -136 |
| en-ca | 92.9 | 91.2 | 1.7 | 53 | 72 | -19 |
| en-cs | 99.3 | 93.7 | 5.6 | 4 | 51 | -47 |
| en-da | 99.6 | 92.9 | 6.7 | 2 | 59 | -57 |
| en-de | 96.6 | 93.9 | 2.8 | 26 | 53 | -27 |
| en-el | 96.0 | 91.8 | 4.2 | 16 | 53 | -37 |
| en-es | 81.9 | 91.3 | -9.4 | 105 | 53 | 52 |
| en-et | 95.1 | 95.4 | -0.2 | 30 | 30 | 0 |
| en-fi | 93.5 | 92.4 | 1.1 | 37 | 51 | -14 |
| en-fr | 90.5 | 89.4 | 1.0 | 86 | 82 | 4 |
| en-hr | 94.0 | 92.4 | 1.6 | 40 | 56 | -16 |
| en-hu | 90.3 | 65.7 | 24.6 | 55 | 282 | -227 |
| en-id | 96.5 | 96.9 | -0.5 | 21 | 17 | 4 |
| en-it | 91.2 | 83.3 | 7.9 | 67 | 127 | -60 |
| en-ko | 94.2 | 94.2 | 0.0 | 32 | 32 | 0 |
| en-nl | 92.7 | 73.3 | 19.4 | 49 | 195 | -146 |
| en-pl | 99.0 | 90.3 | 8.7 | 6 | 70 | -64 |
| en-pt | 94.4 | 86.1 | 8.4 | 50 | 115 | -65 |
| en-ro | 97.2 | 94.0 | 3.2 | 18 | 35 | -17 |
| en-ru | 97.1 | 96.6 | 0.5 | 19 | 21 | -2 |
| en-sl | 95.7 | 91.2 | 4.5 | 24 | 60 | -36 |
| en-sv | 95.8 | 94.1 | 1.8 | 27 | 40 | -13 |
| en-tr | 96.0 | 89.7 | 6.3 | 30 | 75 | -45 |
| es-en | 96.8 | 91.0 | 5.7 | 19 | 61 | -42 |
| et-en | 97.4 | 94.5 | 2.9 | 17 | 42 | -25 |
| fi-en | 99.5 | 94.4 | 5.1 | 2 | 46 | -44 |
| fr-en | 99.8 | 97.5 | 2.4 | 1 | 22 | -21 |
| hr-en | 98.0 | 91.8 | 6.2 | 12 | 69 | -57 |
| hu-en | 91.4 | 83.8 | 7.6 | 59 | 143 | -84 |
| id-en | 98.5 | 94.8 | 3.7 | 11 | 42 | -31 |
| it-en | 93.2 | 88.7 | 4.5 | 47 | 93 | -46 |
| ja-en | 95.7 | 91.7 | 4.1 | 28 | 90 | -62 |
| ko-en | 88.5 | 85.5 | 3.0 | 72 | 121 | -49 |
| lt-en | 94.6 | 95.2 | -0.6 | 42 | 41 | 1 |
| lv-en | 94.5 | 95.0 | -0.5 | 40 | 38 | 2 |
| nl-en | 97.6 | 72.4 | 25.2 | 16 | 287 | -271 |
| nl-en new | 97.6 | 98.0 | -0.4 | 16 | 15 | 1 |
| pl-en | 98.2 | 85.1 | 13.2 | 12 | 140 | -128 |
| pt-en | 97.6 | 91.9 | 5.7 | 16 | 85 | -69 |
| ro-en | 98.9 | 92.1 | 6.9 | 6 | 60 | -54 |
| ru-en | 91.5 | 89.8 | 1.7 | 59 | 81 | -22 |
| sl-en | 95.2 | 86.5 | 8.7 | 28 | 106 | -78 |
| sr-en | 90.7 | 90.2 | 0.5 | 65 | 81 | -16 |
| sv-en | 98.2 | 97.3 | 1.0 | 12 | 25 | -13 |
| tr-en | 87.6 | 86.4 | 1.2 | 85 | 112 | -27 |
| uk-en | 96.5 | 96.3 | 0.2 | 25 | 30 | -5 |
| vi-en | 93.5 | 90.5 | 3.0 | 57 | 100 | -43 |
| zh-en | 95.7 | 92.2 | 3.5 | 26 | 73 | -47 |

NOTE: the test set is likely to be included in the trainings of all the models. So, it is not 100% clear if a model not showing high differential on the third column is only because the sentences were included in the training.

**Comment by eu9ene** [comment]
@ZJaume did I get it right that you just retrained Dutch using our latest pipeline, and it fixed the problem? Or did you do something extra? I remember you've found a bad dataset with short phrases.

**Comment by evgenyrp** [comment]
Just to summarize Jaume's findings, here are the models that have issues with short sentences and should be retrained. The higher the number, the worse it is.

lang | avg % diff
-- | --
en-el | 3.5
en-fr | 3.5
en-fa | 3.9
sl-en | 3.9
en-tr | 7
it-en | 7.5
nb-en | 7.9
nn-en | 8.1
en-bg | 8.5
en-sk | 9.1
en-fi | 11.3
pt-en | 12.5
en-sl | 13.2
bs-en | 15.7
en-da | 15.7
pl-en | 15.8
en-pl | 16.3
en-lt | 29.4
en-pt | 32.7
en-it | 35.3
en-nl | 36.3
en-hu | 48.9

**Comment by evgenyrp** [comment]
We've found that English to Ukrainian gives some bad results on some short examples:
```
hello -> Привіт, (extra comma)
potato -> Картопляна картопля (two times potato)
```

It has avg % diff 2.4, on the border of the mentioned 3. @ZJaume does your analysis cover similar cases? Do you think it means the threshold for bad models is lower than 3?

**Comment by evgenyrp** [comment]
> Given that "just" retraining solved the problem for Dutch, is there anything pending in this issue? (I am tracking [#891](https://github.com/mozilla/translations/issues/891) for the languages that will be retrained)

The plan is to "just retrain and see" for now. I've already started retraining for the languages mentioned in #891. @ZJaume is also working on integrating this metric into the pipeline.

**Comment by DavidVentura** [comment]
I've gotten the files off the repo, at the linked commit.

```
firefox-translations-models/models/base-memory/iten$ git log | head
commit a06d4724eb95d7452f9251cf2cc4ca2706636d74
Author: Evgeny Pavlov <epavlov@mozilla.com>
Date:   Thu Sep 18 06:27:10 2025 -0700

    Add en-zh and en-ja finetuned on 10M LLM generated corpus (#262)

    * Add a model finetuned on 10M LLM generated corpus

    * Add en-ja

firefox-translations-models/models/base-memory/iten$ md5sum *
717abbc8e77c178a1bd01dc7e33a02ec  lex.50.50.iten.s2t.bin.gz
703687e476d27ca963e766af06b2bf51  metadata.json
8246768676e29d2daf54b433be4217a3  model.iten.intgemm.alphas.bin.gz
9baa4edeb3acbf9e573f809bccb30627  vocab.iten.spm.gz

firefox-translations-models/models/base-memory/iten$ git log .
commit 69892e2ee8c42b894fb430c80b36cf68eabb27a3
Author: Evgeny Pavlov <epavlov@mozilla.com>
Date:   Wed Sep 17 12:43:48 2025 -0700

    Retrain high resource part 1 (#263)

    * Add de, bg, et, fi to en

    * Update evaluation results [skip ci]

    * Add nl-en

    * Update evaluation results [skip ci]

    * Add it-en, pt-en

    * Update evaluation results [skip ci]

    * Add cs-en

    * Update evaluation results [skip ci]

    * Add fr-en, es-en

    * Update evaluation results [skip ci]

    * Update evaluation results [skip ci]

    ---------

```

uncompressed hashes 
```
$ md5sum *
7c7bc0267fa1ae10256973bb6ebab4e3  lex.50.50.iten.s2t.bin
703687e476d27ca963e766af06b2bf51  metadata.json
8461122888048fa2484f604f6caaf4bb  model.iten.intgemm.alphas.bin
17e9cb5378bdd60bbd4754a1b59cfb97  vocab.iten.spm
```

maybe these are not the right files?

**Comment by evgenyrp** [comment]
The files are correct; maybe you're running them differently? I think the source of truth for us is `about:translations` in Firefox. Please check if you can reproduce the issue there. There's still room for some randomization based on hardware etc. It might be a one-off thing. If you see a lot of those, something is wrong. Additionally, we have shortlisting disabled due to implementation issues on the Firefox side.

**Comment by DavidVentura** [comment]
This is my config for bergamot

```
models:
  - $dataPath/${languageFiles.model.first}
vocabs:
  - $dataPath/${languageFiles.srcVocab.first}
  - $dataPath/${languageFiles.tgtVocab.first}
shortlist:
    - $dataPath/${languageFiles.lex.first}
    - false
beam-size: 1
normalize: 1.0
word-penalty: 0
max-length-break: 128
mini-batch-words: 1024
max-length-factor: 2.0
skip-cost: true
cpu-threads: 1
quiet: true
quiet-translation: true
gemm-precision: int8shiftAlphaAll
alignment: soft
```

without shortlist I noticed a bit higher quality  (eg: japanese) but quite a bit slower (~2x)

**Comment by DavidVentura** [comment]
@evgenyrp it->en _without_ shortlists does indeed translate properly (no leading ` -`), thank you guys!

**Comment by DavidVentura** [comment]
is shortlisting enabled/fixed in Firefox nowadays?

**Comment by marco-c** [comment]
@DavidVentura it is still disabled.
